src/v/storage/tests/batch_consumer_utils_test.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "bytes/iostream.h"
+#include "model/tests/random_batch.h"
+#include "storage/batch_consumer_utils.h"
+#include "storage/directories.h"
+#include "storage/log_manager.h"
+#include "storage/parser.h"
+#include "storage/segment_reader.h"
+#include "storage/tests/utils/disk_log_builder.h"
+#include "test_utils/tmp_dir.h"
+
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/util/defer.hh>
+
+#include <memory>
+
+struct round_robin_consumer : storage::batch_consumer {
+    consume_result
+    accept_batch_start(const model::record_batch_header& h) const override {
+        if (expect_close) {
+            if (accept_cnt == 0) {
+                accept_cnt++;
+                return consume_result::stop_parser;
+            } else {
+                BOOST_FAIL("Unexpected accept_batch_start");
+            }
+        }
+        if ((static_cast<size_t>(h.base_offset()) % 2) == side) {
+            accept_cnt++;
+            return consume_result::accept_batch;
+        }
+        return consume_result::skip_batch;
+    }
+
+    void consume_batch_start(
+      model::record_batch_header hdr, size_t, size_t) override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected consume_batch_start");
+        }
+        consumed->push_back(hdr);
+    }
+
+    void
+    skip_batch_start(model::record_batch_header hdr, size_t, size_t) override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected skip_batch_start");
+        }
+        skipped->push_back(hdr);
+    }
+
+    void consume_records(iobuf&&) override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected consume_records");
+        }
+    }
+
+    ss::future<stop_parser> consume_batch_end() override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected consume_batch_end");
+        }
+        if (consumed->size() < size_limit) {
+            co_return stop_parser::no;
+        }
+        co_return stop_parser::yes;
+    }
+
+    void print(std::ostream& o) const override { o << "test_consumer"; }
+
+    size_t side{0};
+    size_t size_limit{0};
+    std::deque<model::record_batch_header>* consumed;
+    std::deque<model::record_batch_header>* skipped;
+    // If this is true any call to consume_batch_* or skip_batch_* will
+    // trigger test failure.
+    bool expect_close{false};
+    mutable size_t accept_cnt{0};
+};
+
+SEASTAR_THREAD_TEST_CASE(test_chained_consumers_round_robin) {
```

Comment:
Could we add more test coverages where there are mismatches in skipping and stopping between consumers? It may be difficult with the round robin consumer, but maybe we should have tests that exercise consumers overlapping randomly. Maybe it'd be useful to have a consumer that just stores headers in its own container within a specific offset range and type filter? Then a test could generate overlapping or non-overlapping ranges and validate each resulting container

Reply:
This test suite provides full coverage as far as I can see.

Reply:
I don't think it covers:
- cases where all of the consumers are stopped  early or skipped
- cases where any or all of the consumers hit an exception

Reply:
done